File: frontend/src/types/snapshot.ts
/**
 * Types partagés pour le snapshot SSE.
 *
 * Conformes au schéma défini dans PROJECT.md § "Architecture temps réel"
 * et aux migrations SQLite (backend/migrations/0001_init.sql).
 *
 * Ces types sont actuellement maintenus à la main. Quand le backend
 * grossira, on pourra générer le TS depuis Rust avec ts-rs ou similaire.
 */

import type { DuckColor } from './duck'

/** Un palier (Apéro, Dîner, etc.) tel qu'exposé par le backend. */
export type Phase = {
  id: number
  phase_order: number
  name: string
  /** ISO timestamp. Le calcul current/visible se fait via target_time <= NOW() */
  target_time: string
  /** ISO timestamp si déjà déclenché (peut être forcé par un témoin). */
  triggered_at: string | null
  is_final_reveal: boolean
}

/** Un média photo dans le snapshot. */
export type Media = {
  id: string
  user_id: string
  filename: string
  thumb_filename: string
  posted_at: string
  hidden: boolean
  /** Enrichissements pour l'UI — joints côté backend pour éviter un round-trip */
  user_pseudo?: string
  user_color?: DuckColor
}

export type SnapshotCounts = {
  total_users: number
  total_posts: number
  posts_visible: number
  posts_pending: number
}

/**
 * État complet poussé par le backend via SSE.
 * Pas de delta : à chaque changement, le client remplace son état par
 * ce snapshot (cf. PROJECT.md § "SSE avec snapshot complet").
 */
export type Snapshot = {
  /** ISO timestamp serveur — utile pour calculer le drift d'horloge client. */
  server_time: string
  phase_current: Phase
  phase_visible: Phase | null
  phases_all: Phase[]
  media_visible: Media[]
  /** Médias récents pour Mare TV (palier en cours, non encore visibles). */
  media_recent_for_tv: Media[]
  counts: SnapshotCounts
}

File: backend/src/models/snapshot.rs
use serde::Serialize;

use super::{Media, Phase};

/// Snapshot complet de l'état de la mare. Le client remplace son store
/// local par ce snapshot à chaque changement (pas de delta — voir
/// PROJECT.md § "Architecture temps réel").
#[derive(Debug, Clone, Serialize)]
pub struct Snapshot {
    /// ISO 8601 — heure du serveur, utile pour mesurer le drift client.
    pub server_time: String,
    /// Palier en cours (le dernier dont target_time <= NOW). None si
    /// aucun palier n'a encore démarré.
    pub phase_current: Option<Phase>,
    /// Palier visible dans la galerie (le précédent du current). None
    /// quand on est sur le premier palier ou que phase_current est None.
    pub phase_visible: Option<Phase>,
    /// Tous les paliers (pour orchestration témoins).
    pub phases_all: Vec<Phase>,
    /// Médias visibles dans la galerie (postés dans la fenêtre du
    /// palier visible).
    pub media_visible: Vec<Media>,
    /// Médias récents pour Mare TV (palier en cours, pas encore visibles
    /// dans la galerie publique).
    pub media_recent_for_tv: Vec<Media>,
    pub counts: SnapshotCounts,
}

#[derive(Debug, Clone, Serialize)]
pub struct SnapshotCounts {
    pub total_users: i64,
    pub total_posts: i64,
    pub posts_visible: i64,
    pub posts_pending: i64,
}
